solarforecastarbiter/reference_forecasts/main.py

```diff
@@ -61,18 +70,179 @@ def run(site, model, init_time, start, end):
+    (*index = False*) is supported.
+
+    The persistence *window* is the time over which the persistence
+    quantity (irradiance, power, clear sky index, or power index) is
+    averaged over. The persistence window is automatically determined
+    based on the type of persistence forecast desired:
+
+      * Intraday persistence forecasts: Maximum of observation interval
+        length, forecast interval length, and forecast lead time to
+        start. No longer than 1 hour.
+      * Day ahead forecasts: Equal to forecast interval length.
+
+    Users that would like more flexibility may use the lower-level
+    functions in
+    :py:mod:`solarforecastarbiter.reference_forecasts.persistence`.
+
+    Parameters
+    ----------
+    observation : datamodel.Observation
+        The metadata of the observation to be used to create the
+        forecast.
+    forecast : datamodel.Forecast
+        The metadata of the desired forecast.
+    issue_time : pd.Timestamp
+        Issue time of the forecast run.
+    index : bool, default False
+        If False, use persistence of observed value. If True, use
+        persistence of clear sky or AC power index.
+
+    Returns
+    -------
+    forecast : pd.Series
+        Forecast conforms to the metadata specified by the *forecast*
+        argument.
+
+    Raises
+    ------
+    ValueError
+        If forecast and issue_time are incompatible.
+    ValueError
+        If forecast.run_length >= 1 day and index=True.
+    ValueError
+        If instantaneous forecast and instantaneous observation interval
+        lengths do not match.
+    ValueError
+        If average observations are used to make instantaneous forecast.
+    """
+    forecast_start, forecast_end = get_forecast_start_end(forecast, issue_time)
+
+    # intra day persistence and "day ahead" persistence require
+    # fairly different parameters.
+    # is this a sufficiently robust way to distinguish?
+    intraday = forecast.run_length < pd.Timedelta('1d')
+    if intraday:
+        # time window over which observation data will be used to create
+        # persistence forecast.
+        # assumes observation.interval_length <= 1h
+        window = min(
+            pd.Timedelta('1h'),
+            max(observation.interval_length, forecast.interval_length,
+                forecast.lead_time_to_start, forecast.run_length)
+        )
+        data_end = issue_time
+        data_start = data_end - window
+    else:
+        # day ahead persistence: tomorrow's forecast is equal to yesterday's
+        # observations. So, forecast always uses obs > 24 hr old at each valid
+        # time. Smarter approach might be to use today's observations up
+        # until issue_time, and use yesterday's observations for issue_time
+        # until end of day. So, forecast *never* uses obs > 24 hr old at each
+        # valid time. Arguably too much for a reference forecast.
+        data_end = issue_time.floor('1d')
```

Comment:
Here we're saying that a dayahead forecast is based on the midnight-to-midnight period preceding the issue time. What if the forecast start is noon, and run_length 24 hours?  The forecast would use the index at midnight yesterday for the first hour 12:00 to 1:00pm.  Seems to me that the forecast should use a 24 hour period but that hour-of-day alignment is required.

Reply:
Good catch, this would do the wrong thing in that scenario. I suggest that for now we don't support forecasts that are not midnight-to-midnight.